### PR TITLE
core(image-elements): cache naturalSize results

### PR DESCRIPTION
**Pull request by paulirish**
This url is quite slow to run: https://www.selfridges.com/US/en/cat/womens/newin/

On the CLI, the ImageElements gatherer can take 25-30 seconds to run.

Turns out that gatherer finds ~2000 images, however 99% of them are country flag DOM elements with a css background using the exact same sprite image. 

----------

While we have added [that smart only-do-the-50-most-important-network-images check](https://github.com/GoogleChrome/lighthouse/pull/7274)... we don't handle the case of multiple background images pointing to the same asset that's within the top 50.

But now we do. :)

### CODE REVIEW

**Comment by connorjclark** [comment]
how much faster tho

**Comment by paulirish** [comment]
> how much faster tho

25s to 1s on selfridges

File: lighthouse-core/gather/gatherers/image-elements.js [hunk 1]
```diff
@@ -130,18 +130,28 @@ function determineNaturalSize(url) {
 }
 
 class ImageElements extends Gatherer {
+  constructor() {
+    super();
+    /** @type {Map<string, {naturalWidth: number, naturalHeight: number}>} */
+    this._naturalSizeCache = new Map();
+  }
```
Comment (patrickhulce) [inline lighthouse-core/gather/gatherers/image-elements.js]:
```suggestion
  }
  
```

**Comment by brendankenny** [comment]
> 25s to 1s on selfridges

that doesn't seem possible, fwiw, or the protocol timeout is broken :)

https://github.com/GoogleChrome/lighthouse/blob/6d56a707fe88f77bacbfccf13d6051b5992876dd/lighthouse-core/gather/gatherers/image-elements.js#L141-L142

I get like 8s -> 250ms, which is still 👍

**Comment by brendankenny** [comment]
> > 25s to 1s on selfridges
> 
> that doesn't seem possible, fwiw, or the protocol timeout is broken :)

oh, I take it back. It's because they all share a network record so we just repeatedly fetch the same images 2000 times? Dear lord, yeah, good find.

**Pull request merged**